### ISSUE TITLE
Additional details

### ISSUE THREAD

**Issue by travisdowns** [gregoral/SIMDeefective#1]
I could run this on more hardware, but do you have some details about what test to run and what fails specifically?

Have you managed to reproduce it outside of .NET?

**Comment by gregoral** [comment]
Hi, the code essentially runs a single test in a tight loop.
At the moment all tests calculate the same value.
The only difference is that there are multiple variants ( "simple", loop unrolling, multiple independent registers, ... ).
All vector tests fail in under 30 seconds on 4-th gen processors but not on newer generations.

The "core" of the loop contains only three SIMD instructions in a tight loop:
- Vector.Read
- Vector.Equals
- Vector.Subtract

I conditionally execute non SIMD code to align memory address before entering the vector loop.
But this seems to have no effect other than performance.

**Comment by travisdowns** [comment]
Maybe also compare the .NET platform versions? It seems plausible to me that it is related to e.g., the .NET JIT, perhaps during the transition between different levels of compilation or interpreter and compiler (assuming .NET works like this, I'm more familiar with Java) something goes wrong and a register value gets clobbered.

**Comment by gregoral** [comment]
Initially I didn't pay attention.
But the last tests were all done on .NET 6.0.1 which is about a month old.

I'll test with the last version of .NET as well, but .NET seems unlikely cause.
The reason being that the same code run with the same parameters in a loop returns an incorrect result approximately 1 time out of 10.000 calls.

I plan to convert the code to C++ first and test it on Win 7, Win 10 and Linux.

**Comment by travisdowns** [comment]
Hi Gregor,

I wanted to test your code last night and got as far as turning on my old
Haswell box and connecting a monitor but then my kid woke up and that was
the end of that.

Maybe I'll be able to try today.

For what it's worth, you mention you were able to see it 1 time on Windows
10 and 1 time on Skylake, but that you don't think it really occurs there
because you haven't seen it again, but in my opinion "1 time" is enough if
you're sure it happened and the symptom was similar. Perhaps it just
happens much less frequently.


Travis

**Comment by gregoral** [comment]
You are correct, I don't think the issue cannot possibly occur on Windows 10.
I'm 100% certain that the error occurred on Win10 and i5-6500.
But I'm only 90% certain that my code was not at fault back then.
I did refine the code somewhat after that and I just have no repeatable way to demonstrate at the moment.

This is why I want to convert to a much simpler test in C or C++.

The current crop of tests was originally written to evaluate performance of different SIMD variations of the same code.
The code was not written to hunt for a bug.
It just happens to be a good indicator for SIMD problems.

The plan for next week is to get a better test code in C++.
As it turns out all that is needed is to test if the function return the same result 10.000 times ( or more of course ).
If any of the calls return a different result that means an error has occurred during execution.

**Comment by travisdowns** [comment]
Hi Gregor,

Overnight I tried running the test on my 4th gen i7-4790, but I didn't get
any failures after 18,173,952 tests (the value shown as "Total tests").

This is with Windows 10, however: this machine came with Windows 8, but I
recently upgraded, so it is hard for me to test win 7. I did try installing
Windows 7 in a Hyper-V VM on this box, but after hours of attempting it, I
couldn't get it to work (it always hangs during the install process). I'm
going to try a bit more.

Oen theory of what is happening is that some type of interrupt or DPC or
something similar clobbers register values. Normally, when such a thing
occurs, it will save any register it clobbers so the user code isn't any
wider, but there was an issue with AVX: the registers were all
transparently extended from 128 to 256 bits. So existing code that just
saved a xmm and restored the xmm register could clobber the upper 128 bits
with zero, since for AVX code xmm instructions zero the upper 128 bits.
Basically the save/restore code written for SSE doesn't work with the wider
AVX registers.

To work around this, Intel made the SSE instructions not zero the upper
bits, but preserve them (this came at a considerable performance cost).
However, the AVX instructions, ("VEX encoded") which are almost the same as
their SSE counterparts but generally with separate destination register (3
arg vs 2 arg), do zero the upper bits.

So if you have some "rogue" AVX code which runs in an asynchronous context
and the registers aren't saved around it, it could cause this type of
spurious failure. Maybe it would be interesting to have a simpler test
which does a simpler operation (or, perhaps, just examines the register
value repeatedly) to see if indeed it appears as if the upper 128 bits
suddenly disappear.

I would be happy to help write & test this, but unfortunately I cannot
reproduce the issue yet.

Travis
